data-prepper-core/src/test/java/com/amazon/dataprepper/meter/EMFLoggingMeterRegistryTest.java

```diff
@@ -0,0 +1,410 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.meter;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import io.micrometer.core.instrument.Counter;
+import io.micrometer.core.instrument.DistributionSummary;
+import io.micrometer.core.instrument.FunctionCounter;
+import io.micrometer.core.instrument.FunctionTimer;
+import io.micrometer.core.instrument.Gauge;
+import io.micrometer.core.instrument.LongTaskTimer;
+import io.micrometer.core.instrument.Measurement;
+import io.micrometer.core.instrument.Meter;
+import io.micrometer.core.instrument.MockClock;
+import io.micrometer.core.instrument.Statistic;
+import io.micrometer.core.instrument.Tags;
+import io.micrometer.core.instrument.TimeGauge;
+import io.micrometer.core.instrument.Timer;
+import org.junit.jupiter.api.Test;
+import software.amazon.cloudwatchlogs.emf.environment.EnvironmentProvider;
+import software.amazon.cloudwatchlogs.emf.logger.MetricsLogger;
+import software.amazon.cloudwatchlogs.emf.model.DimensionSet;
+import software.amazon.cloudwatchlogs.emf.model.MetricsContext;
+
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static io.micrometer.core.instrument.Meter.Id;
+import static io.micrometer.core.instrument.Meter.Type;
+import static io.micrometer.core.instrument.Meter.Type.DISTRIBUTION_SUMMARY;
+import static io.micrometer.core.instrument.Meter.Type.TIMER;
+import static org.hamcrest.CoreMatchers.equalTo;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+class EMFLoggingMeterRegistryTest {
+    private static final String METER_NAME = "test";
+    private static final String SERVICE_NAME = "DataPrepper";
+    private static final String TEST_TAG_KEY = "testTagKey";
+    private static final String TEST_TAG_VALUE = "testTagValue";
+    private static final TypeReference<Map<String, Object>> MAP_TYPE_REFERENCE = new TypeReference<Map<String, Object>>() {};
+
+    private final ObjectMapper objectMapper = new ObjectMapper();
+    private final MockClock clock = new MockClock();
+    private final EMFLoggingMeterRegistry registry = spy(
+            new EMFLoggingMeterRegistry(
+                    EMFLoggingRegistryConfig.DEFAULT, new EnvironmentProvider().resolveEnvironment().join(), clock)
+    );
+    private final EMFLoggingMeterRegistry.Snapshot registrySnapshot = registry.new Snapshot();
+
+    @Test
+    void snapshotGetMetricName() {
```

Comment:
Ideally, we'd test these by calling `publish`. Would that be a heavy lift?

Reply:
Do you mean most test cases on snapshot should be tested by registry::publish? The overheads in publish are (1) it aggregates all MetricsLoggers into a list. (2) I could not find a way to mock or spy returned instance method MetricsLogger::flush(). Thus I took the strategy of using registryMetricsLoggersValid() to test aggregation and registryPublish() to test invocation of all MetricsLogger::flush